air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -104,13 +99,43 @@ fn select_by_path_from_stream<'value>(
     Ok(select_result)
 }
 
+fn select_by_path_from_canon_map_stream<'value>(
+    stream: impl ExactSizeIterator<Item = (&'value JValue, RcSecurityTetraplet)> + 'value,
+    lambda: &NonEmpty<ValueAccessor<'_>>,
+    exec_ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<MapLensResult<'value>> {
+    let stream_size = stream.len();
+    let (idx, body) = exctract_idx(lambda, exec_ctx)?;
+
+    let (value, tetraplet) = lambda_to_execution_error!(stream
+        .peekable()
+        .nth(idx)
+        .ok_or(LambdaError::CanonStreamNotHaveEnoughValues { stream_size, idx }))?;
+
+    // csm.$.key.[0] case
+    let select_result = if body.is_empty() {
+        let result = Cow::Borrowed(value);
+        MapLensResult::from_cow(result, tetraplet)
+    } else {
+        // csm.$.key.[0].attribute case
+        let result = select_by_path_from_scalar(value, body.iter(), exec_ctx)?;
+
+        let json_path_suffix = body.iter().fold("".to_string(), |acc, va| acc + "." + &va.to_string());
```

Comment:
BTW, the lambda expression can be rewritten with `format!`, and it will be more efficient than the string addition.

Reply:
According with my research format! won't reduce the number of intermediate String results but Vec::join() will.
```
body.iter().map(ToString::to_string).collect::<Vec<_>>().join(".");
```

Reply:
First, `format!` will store everything into one buffer (yep, recursively).  You implementation allocates a new string for each element.

Second, `format!` is idiomatic way of concatenating string, especially if some of them are static: https://rust-unofficial.github.io/patterns/idioms/concat-format.html

And I was talking about the lambda expression, not whole loop.  BTW, you should use [`Itertools::format`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.format) with `format_args!`.